src/integration-tests/bash/krun.sh

```diff
@@ -0,0 +1,276 @@
+  kubectl delete -n ${NAMESPACE} cm ${PODNAME}-cm --ignore-not-found > /dev/null 2>&1
+}
+
+delete_pod_and_cm
+
+# create cm that contains any files the user specified on the command line:
+
+if [ ! -z "$CMFILES" ]; then
+  kubectl create cm ${PODNAME}-cm -n ${NAMESPACE} ${CMFILES} > $TEMPFILE 2>&1
+  EXITCODE=$?
+  if [ $EXITCODE -ne 0 ]; then
+    echo "Error: kubectl create cm failed."
+    # Since EXITCODE is non-zero, the script will skip 
+    # doing more stuff and cat the contents of $TEMPFILE below.
+  fi
+fi
+
+# run teh command, honoring any mounts specified on the command line
+
+[ $EXITCODE -eq 0 ] && kubectl run -it --rm --restart=Never --tty --image=${IMAGE} ${PODNAME} -n ${NAMESPACE} --overrides "
+{
+  \"spec\": {
+    \"hostNetwork\": true,
+    \"containers\":[
+      {
+        \"command\": [\"bash\"],
+        \"args\": [\"-c\",\"${COMMAND} ; echo -n {{EXITCODE=\$?=EXITCODE}} ; echo -n {{ID=\`id\`=ID}} ; echo -n {{PWD=\`pwd\`=PWD}}\"],
+        \"stdin\": true,
+        \"tty\": true,
+        \"name\": \"${PODNAME}\",
+        \"image\": \"${IMAGE}\",
+        \"imagePullPolicy\": \"${IMAGEPULLPOLICY}\",
+        \"volumeMounts\": [
+          ${VOL_MOUNTS}
+        ]
+      }
+    ],
+    \"imagePullSecrets\": [
+      ${IMAGEPULLSECRETS}
+    ],
+    \"volumes\": [
+      ${VOLS}
+    ]
+  }
+}
+" > $TEMPFILE 2>&1 &
+
+# If the background task doesn't complete within MAXSECS
+# then report a timeout and try kubectl describe the pod.
+
+STARTSECS=$SECONDS
+while [ $EXITCODE -eq 0 ]
+do
+  JOBS="`jobs -pr`"
+  [ "$JOBS" = "" ] && break
+  if [ $((SECONDS - STARTSECS)) -gt $MAXSECS ]; then
+    EXITCODE=98
+    echo "Error: Commmand timed out after $MAXSECS seconds."
+    kubectl describe -n ${NAMESPACE} pod ${PODNAME} 
+    break
+  fi
+  sleep 0.1
+done
+
+# If background task's out file doesn't contain '{{EXITCODE=0=EXITCODE}}', 
+# then it didn't finish or it failed...
+
+if [ $EXITCODE -eq 0 ]; then
+  if [ ! "`grep -c '{{EXITCODE=0=EXITCODE}}' $TEMPFILE`" -eq 1 ]; then
+    EXITCODE=99
+  fi
+fi
+
+# Delete pod in case it timed out and is still running, plus delete
+# the config map of files (if any).  This should also complete the
+# background task (which waits on the pod to finish running).
+
+delete_pod_and_cm
+
+# Show output from pod (or from failing 'kubectl create cm' command above)
+
+if [ $EXITCODE -eq 0 ]; then
+  IDFILTER="{{ID=.*=ID}}"
+  PWDFILTER="{{PWD=.*=PWD}}"
+else
+  # Don't filter out ID and PWD - leave it in for debugging purposes...
+  IDFILTER="{{DNE=DNE}}"
+  PWDFILTER="{{DNE=DNE}}"
+fi
+
+cat $TEMPFILE | sed "s/{{EXITCODE=0=EXITCODE}}//" \
+              | sed "s/$IDFILTER//" \
+              | sed "s/$PWDFILTER//" \
+              | grep -v "Unable to use a TTY - input is not a terminal or the right kind of file" \
+              | grep -v "If you don't see a command prompt, try pressing enter." \
+              | grep -v "Error attaching, falling back to logs:"
+
```

Comment:
FYI:   When I wrote/tested this script, I only tried it out on 1.10.11 (client and server).   I don't know if the above 'grep -v' commands are sufficient to filter out the stderr/stdout extraneous cruft on all version of k8s.   This means a krun.sh output might unexpectedly  include some unexpected extra output sometimes on some versions -- which in turn would mean the 'base64' copy trick won't work.

Reply:
If you do see such cruft, I have an idea about a potentially more generic way to filter it out (print sentinel values at start and end of a run, and filter out everything before and after these sentinel values).

Reply:
Ok.

Reply:
Ok